lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -183,6 +198,24 @@ def __init__(self, name: AnyStr, type: ParamType, required: bool = False):
         self.type = type
         self.required = required
 
+    def invalid(self, json: Dict[AnyStr, Any]) -> bool:
+        """
+        ok Check whether the expression of this parameter in the JSON document
+        is invalid.
+
+        If the parameter is required and missing (or null), then it's invalid;
+        if the parameter is not required and specified as null, it's invalid.
+
+        Args:
+            json: The client JSON document being validated.
+
+        Returns:
+            True if unacceptable
+        """
+        return (self.required and self.name not in json) or (
+            self.name in json and not json[self.name]
```

Comment:
It's disappointing that this expression checks the list twice; how about this instead?
```
        return self.required if self.name not in json else not json[self.name]
```
or this (with one fewer `not`):
```
        return not json[self.name] if self.name in json else self.required
```

Reply:
I went for easy to read rather than noodling out ultimate efficiency, but I'll give this some thought. I'm going to ride out today on my index template refactoring anyway, so I'll get back to this on Monday.